### PR TITLE
Fix clang-tidy warnings

### PR DESCRIPTION
**Pull request by masterleinad**
We don't enforce `clang-tidy` warnings anymore and there are multiple complaints.
This pull request enforces warnings by turning them into errors and fixes the current warnings.

### CODE REVIEW

File: benchmarks/dbscan/ArborX_DBSCANVerification.hpp [hunk 1]
```diff
@@ -306,7 +306,7 @@ bool verifyDBSCAN(ExecutionSpace exec_space, Primitives const &primitives,
   ARBORX_ASSERT(eps > 0);
   ARBORX_ASSERT(core_min_size >= 2);
 
-  Points points{primitives};
+  Points const &points{primitives};
```
Comment (aprokop) [inline benchmarks/dbscan/ArborX_DBSCANVerification.hpp:309]: What warnings are here? This makes little sense to me. I would not want to change these.

File: test/tstMinimumSpanningTreeGoldenTest.cpp [hunk 2]
```diff
@@ -52,8 +52,8 @@ auto parsePointsFromCSVFile(std::string const &filename)
     Tokenizer tok(line);
     auto first = tok.begin();
     auto const last = tok.end();
-    points.emplace_back(ArborX::Point{std::stof(*first++), std::stof(*first++),
-                                      std::stof(*first++)});
+    points.emplace_back(std::stof(*first++), std::stof(*first++),
+                        std::stof(*first++));
```
Comment (aprokop) [inline test/tstMinimumSpanningTreeGoldenTest.cpp:56]: What happened her?

Comment (masterleinad) [reply test/tstMinimumSpanningTreeGoldenTest.cpp:56]: `emplace_back` takes the arguments of the constructor. The changes here avoid calling the move constructor.

File: benchmarks/dbscan/ArborX_DBSCANVerification.hpp (hunk 1 above)
Comment (masterleinad) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp:309]:
```
 warning: local copy 'points' of the variable 'primitives' is never modified; consider avoiding the copy [performance-unnecessary-copy-initialization]
```
We could just `NOLINT` here.

File: benchmarks/distributed_tree_driver/distributed_tree_driver.cpp [hunk 3]
```diff
@@ -219,7 +219,7 @@ struct ArborX::AccessTraits<NearestNeighborsSearches<DeviceType>,
 namespace bpo = boost::program_options;
 
 template <class NO>
-int main_(std::vector<std::string> const &args, const MPI_Comm comm)
+int main_(std::vector<std::string> const &args, MPI_Comm const comm)
```
Comment (masterleinad) [inline benchmarks/distributed_tree_driver/distributed_tree_driver.cpp:222]: This and some other similar changes are related to using a newer clang-format version but I didn't bother to separate since I don't have `clang-format-14` anymore.

File: benchmarks/dbscan/ArborX_DBSCANVerification.hpp [hunk 4]
```diff
@@ -254,12 +254,12 @@ bool verifyClustersAreUnique(ExecutionSpace const &exec_space,
   }
   if (cluster_sets.size() != num_unique_cluster_indices)
   {
-    std::cerr << "Number of components does not match" << std::endl;
+    std::cerr << "Number of components does not match" << '\n';
```
Comment (dalg24) [inline benchmarks/dbscan/ArborX_DBSCANVerification.hpp:257]: I'd expect we intended to flush here but if we didn't then the `\n` belongs to arrays and should not be appended as a single char.

Comment (masterleinad) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp]: `std::cerr` isn't buffered so we don't need to flush.

Comment (dalg24) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp]: I did not know that :)

File: benchmarks/dbscan/ArborX_DBSCANVerification.hpp (hunk 1 above)
Comment (aprokop) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp]: Would using `Points const points{primitives};` work? If yes, I would prefer that to `NOLINT`.

Comment (masterleinad) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp]: No, the point is that the copy could be avoided. `const` doesn't make a difference.

**Comment by masterleinad** [comment]
The test failures are due to "No test on device".

File: benchmarks/dbscan/ArborX_DBSCANVerification.hpp (hunk 1 above)
Comment (aprokop) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp]: Can we move `NOLINT` to inside `AccessTraits` somehow?

Comment (masterleinad) [reply benchmarks/dbscan/ArborX_DBSCANVerification.hpp]: I don't think so. The problem is that `Points` and `Primitives` could be the same type if the latter is a `Kokkos::View`. Another workaround would be
```C++
std::conditional_t<std::is_same_v<Predicates, UserPredicates>, Predicates const&, Predicates> predicates{user_predicates};
```
but I'd prefer NOLINT.

File: .clang-tidy [hunk 5]
```diff
@@ -1,5 +1,5 @@
 ---
-Checks:          'clang-diagnostic-*,clang-analyzer-*,-*,performance-*,-performance-inefficient-string-concatenation,mpi-*,modernize-*,-modernize-pass-by-value,-modernize-use-trailing-return-type,-modernize-avoid-c-arrays,-modernize-use-nodiscard,-modernize-concat-nested-namespaces,readability-*,-readability-braces-around-statements,-readability-named-parameter,-readability-magic-numbers,-readability-uppercase-literal-suffix,-readability-identifier-length,-readability-function-cognitive-complexity,-readability-redundant-string-cstr'
+Checks:          'clang-diagnostic-*,clang-analyzer-*,-*,performance-*,-performance-noexcept-swap,-performance-inefficient-string-concatenation,mpi-*,modernize-*,-modernize-pass-by-value,-modernize-use-trailing-return-type,-modernize-avoid-c-arrays,-modernize-use-nodiscard,-modernize-concat-nested-namespaces,readability-*,-readability-braces-around-statements,-readability-named-parameter,-readability-magic-numbers,-readability-uppercase-literal-suffix,-readability-identifier-length,-readability-function-cognitive-complexity,-readability-redundant-string-cstr,-readability-inconsistent-declaration-parameter-name'
```
Comment (aprokop) [inline .clang-tidy:2]: Why were `noexcept-swap` and `inconsistent-declaration-parameter-name` disabled? With the latter one, does it mean that there is no need to do the change in `ArborX_DetailDistributedTreeImpl.hpp`?

File: src/ArborX_DistributedTree.hpp [hunk 6]
```diff
@@ -50,7 +50,6 @@ class DistributedTreeBase
   using bounding_volume_type = BoundingVolume;
   using value_type = typename BottomTree::value_type;
 
-public:
```
Comment (aprokop) [inline src/ArborX_DistributedTree.hpp:53]: Why was this removed?

File: .clang-tidy (hunk 5 above)
Comment (aprokop) [reply .clang-tidy:2]: Removing this checks produced no warnings (at least, for `Serial` build).

Comment (masterleinad) [reply .clang-tidy:2]: > Reenabling these checks produced no warnings (at least, for Serial build).

Did you try a recent version of `clang-tidy`? I was running with `17.0.6`.

> Why were noexcept-swap and inconsistent-declaration-parameter-name disabled? 

Because they are noisy and don't add much value. `noexcept-swap` complains if our `swap` implementation is not `noexcept` but we probably don't care. `inconsistent-declaration-parameter-name` suggests a lot of changes that we could make if we care about naming our parameters consistently but it would be better to create another pull request for that.

> With the latter one, does it mean that there is no need to do the change in ArborX_DetailDistributedTreeImpl.hpp?

Yes, it seems I forgot to remove that one.

Comment (aprokop) [reply .clang-tidy:2]: > Did you try a recent version of `clang-tidy`? I was running with `17.0.6`.

Was running with 14.

> `noexcept-swap` complains if our `swap` implementation is not `noexcept` but we probably don't care. ``

Why don't we instead make our `swap` `noexcept`?

> `inconsistent-declaration-parameter-name` suggests a lot of changes that we could make if we care about naming our parameters consistently but it would be better to create another pull request for that.

Weird. The option is present in 14, so I should have seen a bunch of warnings. But I don't. It may be a good idea to at least take a look at it.

Comment (masterleinad) [reply .clang-tidy:2]: > Why don't we instead make our `swap` `noexcept`?

Because that would be a lie. Our current interface looks like
```C++
template <class T>
KOKKOS_FUNCTION constexpr void
swap(T &a, T &b) noexcept(std::is_nothrow_move_constructible_v<T>
                              &&std::is_nothrow_move_assignable_v<T>);
```
so it's `noexcept` if it can be already. I would rather disable this very specific `check` than change the implementation.

> Weird. The option is present in 14, so I should have seen a bunch of warnings. But I don't. It may be a good idea to at least take a look at it.

If it doesn't show up in the CI (which it probably doesn't if you don't see it), it would be pretty hard to enforce (unless we update to a newer version).

**Comment by masterleinad** [comment]
Retest this please.
